packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol

```diff
@@ -0,0 +1,292 @@
+                        mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x0000001055494e543235365f4f564552464c4f57000000000000000000000000)
+                        mstore(96, 0)
+                        revert(0, 100)
+                    }
+
+                    // Write `totalAmount` to memory
+                    mstore(100, totalAmount)
+
+                    // Load offset to `nestedAssetData[i]`
+                    let nestedAssetDataElementOffset := calldataload(add(nestedAssetDataContentsStart, i))
+
+                    // In order to find the start of the `nestedAssetData[i]` contents, we must add:
+                    // 4 (function selector) 
+                    // + assetDataOffset 
+                    // + 32 (assetData len)
+                    // + 4 (assetProxyId)
+                    // + nestedAssetDataOffset
+                    // + 32 (nestedAssetData len)
+                    // + nestedAssetDataElementOffset
+                    // + 32 (nestedAssetDataElement len)
+                    let nestedAssetDataElementContentsStart := add(assetDataOffset, add(nestedAssetDataOffset, add(nestedAssetDataElementOffset, 104)))
+
+                    // Load length of `nestedAssetData[i]`
+                    let nestedAssetDataElementLenStart := sub(nestedAssetDataElementContentsStart, 32)
+                    let nestedAssetDataElementLen := calldataload(nestedAssetDataElementLenStart)
+
+                    // Revert if the `nestedAssetData` does not contain a 4 byte `assetProxyId`
+                    if lt(nestedAssetDataElementLen, 4) {
+                        // Revert with `Error("LENGTH_GREATER_THAN_3_REQUIRED")`
+                        mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x0000001e4c454e4754485f475245415445525f5448414e5f335f524551554952)
+                        mstore(96, 0x4544000000000000000000000000000000000000000000000000000000000000)
+                        revert(0, 100)
+                    }
+
+                    // Load AssetProxy id
+                    let assetProxyId := and(
+                        calldataload(nestedAssetDataElementContentsStart),
+                        0xffffffff00000000000000000000000000000000000000000000000000000000
+                    )
+
+                    // To lookup a value in a mapping, we load from the storage location keccak256(k, p),
+                    // where k is the key left padded to 32 bytes and p is the storage slot
+                    mstore(132, assetProxyId)
+                    mstore(164, assetProxies_slot)
+                    let assetProxy := sload(keccak256(132, 64))
```

Comment:
One possible optimization: If `assetProxyId` is the same as the previous iteration then we could skip this check. Clients could then optimize their calldata by grouping-by-proxy.